x-pack/plugins/canvas/canvas_plugin_src/functions/types/common.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/**
+ * Utility type for converting a union of types into an intersection.
+ *
+ * This is a bit of "black magic" that will interpret a Union type as an Intersection
+ * type.  This is necessary in this case of distiguishing one collection from
+ * another in `FunctionError` and `FunctionStrings`.
+ */
+// prettier-ignore
+export type UnionToIntersection<U> = 
+  (U extends any ? (k: U) => void : never) extends ((k: infer I) => void) ? I : never;
+
+/**
+ * Utility type: gathers values of a collection as a type for use as a type.
+ */
+export type ValuesOf<T extends any[]> = T[number];
```

Comment:
TS ramp-up (for me) question, what's the difference between this and eg. `Array<T>`?

Reply:
You should only use `Array<T>` for complex type expressions, (and `eslint` will complain if you deviate).  For example: `string[]` and `FooType[]` are preferable to `Array<string>` or `Array<FooType>`, but `Array<{foo: string}>` is preferable in that case.

Reply:
I read it - correctly? - as:
- "You should use Array<T> for complex type expressions only" (ie. not for simple expressions, though I'm not yet clear on what constitutes simple vs complex)
- `string[]` and `FooType[]` are semantically, but not stylistically equivalent to `Array<string>` and `Array<FooType>`, resp.

I tried to trigger the syntax check, entering
```
export type ValuesOf<T extends any[]> = Array<T>;
```
but it didn't balk, maybe it still accepts it or I don't run the right linter. Also, `ValuesOf` and `UnionToIntersection` seem to be unused atm. All these are only curiosities, not sure if they're important at all.

Reply:
I think generics may be an exception...?